### ISSUE TITLE
64 b and array Fragment output

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#902]
doubles ant longs cannot be written as color attachment directly. Even if the device supports them. I assume because `vec3` and `vec4` would take two `Location`, which would prevent interfacing with color attachments. But that feels like API defect; perhaps some extension should figure some healthy way around this?

Not sure how array image color attachments would work. They do not seem to be forbidden in VU at `vkCreateFramebuffer`, but they would take N `Location`s. But at the same time the spec requires that `Location` and `pColorAttachment` index matches.

**Comment by jeffbolznv** [comment]
Yes, array attachments are allowed. Each element of the array gets a consecutive location assigned to it, and this is described (maybe not perfectly clearly) in the "Location Assignment" section.

**Comment by krOoze** [comment]
@jeffbolznv No, that part is clear. Problem is it is incompatible with the "one Location == one attachment" rule.

E.g. If I have Image array[2] as pColorAttachments[0]. It would take two Locations in shader 0 and 1. But it cannot because Location 1 belongs to pColorAttachments[1].

**Comment by jeffbolznv** [comment]
It's the implicit location that is supposed to match up with the pColorAttachment. So you are objecting that it says "code:Location decoration" rather than talking about the implicit location?

**Comment by krOoze** [comment]
What's the difference? If I use explicit Location I still have to follow the location assignment rules (otherwise the data would alias), no?

**Comment by krOoze** [comment]
I am objecting that if I have an Image array `i1[2]` in `pColorAttachments[0]`, and different Image `i2` in `pColorAttachments[1]`, then it is ambiguous if `layout (location = 1) out dvec2 outFrag` refers to Image `i1[1]` or `i2`. Location asignment rules say it must be `i1[1]`, but attachment interface matching says it must be `i2`.

**Comment by krOoze** [comment]
> Image array layers aren't spread across outputs, the only way to select the layers is with the gl_Layer output in the VTG shader.

Thanks. With this I found this in Vulkan spec:

> If the last active vertex processing stage shader entry point’s interface does not include a variable decorated with Layer, then the first layer is used.

It feels though some note to this is missing in the Fragment Output Interface chapter. It is important detail that Framebuffer's (i.e. attachment's) array is completely unrelated to the in-shader array.

<hr />

So what about the >128 b types (e,g, `R64G64B64A64`)? How does that work? I can't decalre those in shader without taking two Locations too, so they cannot be matched to the single attachment. Are they banned (cannot find a VU though), or is there some trick to it too.

**Comment by jeffbolznv** [comment]
Fragment shader outputs can't be 64b per component. I'm not aware of anybody ever having tried to support this.

```
Output values written by a fragment shader must: be declared with either
code:OpTypeFloat or code:OpTypeInt, and a Width of 32.
ifdef::VK_VERSION_1_1,VK_KHR_16bit_storage[]
If pname:storageInputOutput16 is supported, output values written by a
fragment shader can: be also declared with either code:OpTypeFloat or
code:OpTypeInt and a Width of 16.
endif::VK_VERSION_1_1,VK_KHR_16bit_storage[]
```

**Comment by krOoze** [comment]
@jeffbolznv I see. Could be caught earlier at `vkCreateFramebuffer` time with VU then, I think.

**Comment by jeffbolznv** [comment]
I think something like that may already be in the works, for unrelated reasons. Hopefully in the next few weeks.
